compiler/rustc_errors/src/lib.rs

```diff
@@ -332,6 +337,10 @@ struct HandlerInner {
     /// twice.
     emitted_diagnostics: FxHashSet<u128>,
 
+    /// This contains all lint emission information from this compiling session
+    /// with the emission level `Expect`.
+    expected_lint_emissions: Vec<LintEmission>,
```

Comment:
Suggestion: what if we turn this the other way? Say we have a `FxHashMap` (which is a hash map with a faster hashing algorithm that we use throughout the compiler) that represents all the `#[expect]` attributes that need to be fulfilled. It takes a `(LintStackIndex, LintId)` pair as its key and everything that we need to emit the lint (the `#[expect]` attribute's span, the reason, etc) as its value. Every time we emit an `expect`ed lint, we remove that lint's entry from the map, and at the end of compilation, we emit an `unfulfilled_lint_expectation` lint for every remaining entry in the map.

This would require a few changes to how lints are emitted so I don't feel confident to say "it's the way to go", maybe you'll want to wait on someone more qualified to weigh in.

Reply:
This is a very interesting idea that I haven't thought about! I actually think that it won't require any major change to the lint emission. The logic already retrieves the lint level for a specific lint. The implementation would only find away to store which expectation set the level to expect in this concrete stance and mark that one as fulfilled.

I'll try to do some prototyping for this suggestion, as I believe it would be the nicer solution. As log as we can find a way to match the diagnostic to the specific expectation.